### ISSUE TITLE
Changes to mentor blocking behaviour

### ISSUE THREAD

**Issue by iHiD** [exercism/v2-feedback#121]
TL;DR: We will be introducing "Independent Mode" for tracks, where users can choose to progress through the track at a faster pace, without approval, and with limited mentoring. We will also be tweaking the unlocking mechanism for normal usage.

## Quick Recap

The most controversial of the changes in v2 was the decision to base progression on mentor approval. There was an [in-depth discussion](https://github.com/exercism/v2-feedback/issues/36) which was extremely helpful in highlighting the negative tradeoffs of this decision, and which lead us to explore further how some "power-users" use Exercism to learn. Specifically I found the following comment by medwards1771 summed up the problems with this new approach well:

> I like the unstructured old version because it placed trust in users to work at their own pace and progress however they saw fit for themselves.

On the back of this public discussion, the product team spent a lot of time considering the various tradeoffs and working out a compromise that we feel honours the outcome of the research and thought we've put into v2, but also maintains people's ability to work at their own pace in their own way.

### Independent Mode

We are adding Independent Mode to tracks. The usage is for people who want to take "deep dives" into tracks at their own pace without being restricted to waiting for mentors.

When someone starts a track, they can choose to complete it in "Independent mode" or "Standard mode". Standard mode is the default is is the mentored version that is there at the moment. We will discourage independent mode through the UI, without making it feel negative.

Independent mode then works as follows:
- There is no seperation of core/side exercises.
- All exercises are unlocked - there is no unlocking.
- By default, exercises do not receive mentoring. Users can choose to request mentoring on exercises if they want to rather than marking them as completed. Users can only choose to have mentoring on 3 exercises at once. *(Avoids overloading mentors)*
- Exercises completed without mentor approval cannot be publicly published. *(Avoids flooding with solutions of unknown quality)*
- Exercises completed with mentor approval can be publicly published.

### Mentoring changes

- Core exercises will always appear above side exercises in the mentor queues. As core exercises block progression in Standard Mode, but side exercises don't, it's important that core exercises are processed first.
- Exercises in Standard Mode tracks will appear higher in the queues than exercises in "independent mode" tracks.

### Unlocking changes

In Standard Mode, currently side-quests are unlocked when a core exercise is approved by a mentor. There was some concern that people might submit an exercise and then have nothing to do. We are addressing this by changing the behaviour so that that 1-3 side-quests are unlocked when a core exercise is *first submitted* and the rest will unlock once a mentor approves it.

## Summing up

Our aim with these changes is to let people to dig into a track in their own autonomous way while still being able to access the benefits of mentorship, while not overloading mentors or detracting from our core principles of mentor-based learning at a more modest pace. 

We would welcome and feedback you have :)

**Comment by rpottsoh** [comment]
> When someone starts a track, they can choose to complete it in "Independent mode" or "Standard mode". 

Can the student make this choice after starting a track, or are they locked in forever when they first start?  Can they switch to "Independent mode" at a later time if they decided?  Would they be allowed to switch back to "Standard mode" after having progressed a ways in the other mode?  I can see switching back to "Standard" as maybe being messy; dealing with core tracks maybe being unlocked out of order.

Just trying to understand the full scope of this new feature.  Having a choice now I think is a great addition to V2.  Thanks.

**Comment by iHiD** [comment]
Thanks @rpottsoh 

No - the design that a user wouldn't chop and chance between - you make the choice and stick with it. 

If we allow changing, then people will start getting very weird experiences as things they are half way through suddenly become locked, etc. If you're in Independent Mode and you wish you were in normal mode, then really just slowing down will achieve the same thing anyway - you just remain slightly freer.

**Comment by kytrinyx** [comment]
I do think it would be worth considering allowing people to switch from Standard Mode to Independent Mode, but not back.

**Comment by iHiD** [comment]
@kytrinyx Yeah - I'm not against that. I think my gut instinct is to put both modes in, see how they're used, and see whether people ask to be able to switch either way and then reconsider it. However, I'd also be happy with having what you suggest as the starting point and observe from there instead.

**Comment by NobbZ** [comment]
I have about 20 started tracks, how will they be converted? Will I be able to choose? Or will I put in independent mode because this comes closest to how exercism worked? 

There are some tracks I'd like to do in a strictly mentored fashion if at all.

**Comment by iHiD** [comment]
Great question. I think we'll have to ask the question for existing tracks.

**Comment by Insti** [comment]
> Independent Mode
> ...
> By default, exercises do not receive mentoring. 

I strongly disagree with this.
Why are we submitting solutions at all if no-one will see them?



Edit: to clarify I was talking about Independent mode.

**Comment by Insti** [comment]
> We will be introducing "Independent Mode" for tracks, where users can choose to progress through the track at a faster pace

This is great 👍

**Comment by Insti** [comment]
What happens if the mentor backlog becomes big and solutions are not being reviewed and everybody switches to independent mode?

Everyone is submitting solutions that are not visible to anyone other than themselves?

Note: This is not a suggestion to forbid switching!

**Comment by NobbZ** [comment]
>> By default, exercises do not receive mentoring.
>
> I strongly disagree with this.
> Why are we submitting solutions at all if no-one will see them?

This is for the new "independant mode". You can do all the exercises in a single night of hacking session and submit them all at once and then try priotize them in the way you want on the mentoring queue.

And after it has been mentored it can be publicly visible.

Remember, even for the "mentored mode", even mentored and approved submissions are invisible by default. A student has to opt in for every single exercise to make his submission public.

**Comment by ErikSchierboom** [comment]
> A student has to opt in for every single exercise to make his submission public.

Is this correct? If so, I would at least expect the user to be able to choose between making each submission public individually or just have them all be public (which would save the user from having to make the same choice over and over again).

**Comment by NobbZ** [comment]
When a submission was approved by the mentor, I had to confirm the approval. Then a window popped up, asking me if I want to publish it.

I'm not sure about the default in the modal, but the question definitively was there. As far as I remember thats the same modal that also asks you for another comment.

This is based on fading memory about how it was before the data reset, since after the reset progressing, mentoring and therefor approval was literally impossible.

**Comment by Insti** [comment]
> after the reset progressing, mentoring and therefor approval was literally impossible.

👍

**Comment by kytrinyx** [comment]
> I would at least expect the user to be able to choose between making each submission public individually or just have them all be public

@ErikSchierboom That's a good observation, thanks.

> This is based on fading memory about how it was before the data reset, since after the reset progressing, mentoring and therefor approval was literally impossible.

@iHiD and I will be planning the next steps shortly, and we'll get back to you with a plan for—among other things—a new reset.

**Comment by topfunky** [comment]
Is there a current plan for how to find, vet, onboard, and maintain mentors? I’ve build this process a few times (comercially) and it’s a difficult problem.

I’m glad to see that Independent Mode will be supported, which can make it easier for students to participate. The hardest thing as a student on the site (for me) is the lack of community/communication (especially after reaching the more advanced exercises).

**Comment by iHiD** [comment]
> Is there a current plan for how to find, vet, onboard, and maintain mentors? I’ve build this process a few times (comercially) and it’s a difficult problem.

@topfunky We're working on it but would hugely appreciate your insight/feedback. Maybe I could pull you into a Slack conversation with @kytrinyx, @nicolechalmers and myself at some point and pick your brains in depth about it. If you could spare the time?

**Comment by boddhisattva** [comment]
> Exercises completed without mentor approval cannot be publicly published. (Avoids flooding with solutions of unknown quality)

Does this mean that one cannot get feedback/reviews from others who aren't active mentors for a given track ? I mean will the solutions not be visible to other peers and folks(who may not necessarily be official mentors on the track) who could give feedback based on their availability ?

**Comment by iHiD** [comment]
@boddhisattva That's correct.

**Comment by boddhisattva** [comment]
@iHiD Thank you for your prompt response. Can you please explain the rationale behind having this school of thought ? As much as I'd appreciate and welcome feedback from mentors of the track through their valuable time, I'd love to also get feedback from my peers and other folks in the community as and when they'd have time to give me inputs on my solutions even if they aren't necessarily official mentors on Exercism. I understand there may be no easy way to validate the quality of their feedback but another way to look at it also is they could make things easier for the official mentors of a given track through their reviews.

Also, the current way I get insights on how to improve my submitted solutions is by reviewing solutions of some of those who've already completed them. How will the Independent mode affect the viewing of already submitted and now public solutions on Exercism and how will the concept of multiple iterations to a given solution now work in the Independent mode?

**Update**:

In the past, I've also facilitated a meetup where people solved an exercism kata via pair programming (I've written a blog on the experience [here](https://medium.com/@mohnishgj/learnings-from-facilitating-a-ruby-meetup-where-people-solved-an-exercism-kata-via-pair-programming-39c78113b86b) if you'd like to get more context). At such meetups/events in general, if we need to view each others solutions, how could one make sharing our existing solutions between peers more easier if we go by the current way of how the Independent mode will be structured i.e., without them being made publicly available until a mentor has approved a given solution. Will we have to do things like copy our code to gists and then share them with others ?

**Comment by topfunky** [comment]
@iHiD Yes, glad to join a Slack. Email boss@topfunky.com or DM me.

**Comment by NobbZ** [comment]
@boddhisattva 

You will be able to comment on approved by mentors and then published by the student solutions.

**Comment by boddhisattva** [comment]
@NobbZ Thanks for attempting to answer my questions. I'm able to get an idea of what we're planning to do in the Independent mode through yours and @iHiD's comment [here](https://github.com/exercism/v2-feedback/issues/121#issue-265819050) but I'm still not able to get the `why` aspect of this. Also, I'm still trying to better understand how does the concept of iterations and already submitted solutions change with these additional changes, it would be great if one could throw more light on this.

What I like about the current way of using Exercism is it empowers me to:
1. Make my solutions public through each submitted iteration, allowing me to reach out to anyone without necessarily being dependent on only mentors to get the first set of inputs/feedback that could help improve my solution.
2. See the submitted solutions of other folks, sometimes even closely relating to the mistakes their making for the way they've gone about solving a problem especially when I look back and see that I've made similar mistakes in my solutions via different iterations .
3. Sometimes just living those `aha` moments of why I didn't think about doing it this way :)

Doing things this way as I go about solving problems gives me the joy to keep coming back to using Exercism time and again :)

All in all, I'm just another Exercism fan and I'm thankful to all of your efforts to help make this better . 

It's just that as a programmer,  I more closely relate to:
1. Seeing how others are approaching a given problem in general through their journey of submitting iterations and the insightful discussions around each of those iterations. 

2. Living the feeling that it's ok if you've approached a solution in a particular way. IMHO one in general doesn't necessarily have to feel bad about the quality of their solution as it's all part of the learning process of one trying to get better :). I guess it's even ok to submit that kind of a solution, you may end up finding another solution that looks quite similar to yours and then you just end up having a good excuse that could be a conversation starter, something along the lines of `Hey, even I did the same thing that you did as part of your solution :)`.  I guess that once such conversations and connections happen automatically learning becomes even more fun and that's something I look forward to each time I use Exercism. Also this way, a learner to a new language doesn't necessarily have to get bogged down with the feeling that `is it only me` who's not getting it as they go about learning new stuff.

Thanks for allowing me to share my thoughts on this.

**Comment by iHiD** [comment]
@boddhisattva Thank you for your comments and sorry it's taken a bit of time to reply. Both @kytrinyx and I have had very busy periods recently.

- We will have Exercism Teams for situations like the Bootcamp you mention.
- The rational for the changing to mentorship is explained [here](https://github.com/exercism/v2-feedback/blob/master/README.md#a-new-mentorship-model).
- I 100% agree with everything you said about learning from others' solutions. We're not taking away the ability to see others' solutions - in fact we've greatly enhanced it. The situation here is purely around Independent Mode where people are taking fast, deep-dives into code and don't want mentorship or feedback to get in the way of them being able to do lots of exercises. For the 99% case, mentorship, learning from others' solutions and then publishing your own solutions will be at the heart of the workflow.

> I guess that once such conversations and connections happen automatically learning becomes even more fun and that's something I look forward to each time I use Exercism. 

That's exactly the thing we're trying to make happen. Everything you're saying strikes me as the normal use-case for Exercism and that you are exactly the person we have redesigned everything for. 

> Thanks for allowing me to share my thoughts on this.

Thank you for sharing them in such an empathetic way.

**Comment by boddhisattva** [comment]
@iHiD Firstly, I'd like to apologise for a late response from my end too as I've been caught up with a few other things. Thank you for taking the time out to answer my questions :) It's great to know that we could still continue to use the existing features of Exercism in Nextercism. I'm also looking forward to experiencing the Mentorship related changes that come as part of Nextercism. I'm thankful to you, Katrina and others for their continued work on this so that we could have an enhanced learning experience as we go about solving the exercises on the Exercism platform.
